// chatter/dist/util/queue.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Queue = function () {
  function Queue() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _classCallCheck3.default)(this, Queue);

    this.cache = {};
    var _options$onDrain = options.onDrain,
        onDrain = _options$onDrain === undefined ? function (data) {
      return console.log('onDrain', data);
    } : _options$onDrain;

    this.onDrain = onDrain;
  }

  (0, _createClass3.default)(Queue, [{
    key: 'enqueue',
    value: function enqueue(id, data) {
      if (!this.cache[id]) {
        this.cache[id] = {
          queue: []
        };
      }
      var obj = this.cache[id];
      obj.queue.push(data);
      if (!obj.promise) {
        obj.promise = this.drain(id);
      }
      return obj.promise;
    }
  }, {
    key: 'drain',
    value: function drain(id) {
      var _this = this;

      var obj = this.cache[id];
      var next = function next() {
        if (obj.queue.length === 0) {
          obj.promise = null;
          return null;
        }
        return _bluebird2.default.try(function () {
          return _this.onDrain(id, obj.queue.shift());
        }).then(next);
      };
      return next();
    }
  }]);
  return Queue;
}();

exports.default = Queue;
