// chatter/dist/message-handler/command.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommandMessageHandler = undefined;

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports.default = createCommand;

var _delegate = require('./delegate');

var _matcher = require('./matcher');

var _matcher2 = _interopRequireDefault(_matcher);

var _parser = require('./parser');

var _parser2 = _interopRequireDefault(_parser);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function formatCommand() {
  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  // console.log(`formatCommand <${args.join('> <')}>`);
  return args.filter(Boolean).join(' ');
}

var CommandMessageHandler = exports.CommandMessageHandler = function (_DelegatingMessageHan) {
  (0, _inherits3.default)(CommandMessageHandler, _DelegatingMessageHan);

  function CommandMessageHandler() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var children = arguments[1];
    (0, _classCallCheck3.default)(this, CommandMessageHandler);

    var _this = (0, _possibleConstructorReturn3.default)(this, (CommandMessageHandler.__proto__ || (0, _getPrototypeOf2.default)(CommandMessageHandler)).call(this, options, children));

    var name = options.name,
        aliases = options.aliases,
        usage = options.usage,
        description = options.description,
        details = options.details,
        isParent = options.isParent;

    _this.isCommand = true;
    _this.name = name;
    _this.usage = usage;
    _this.description = description;
    _this.details = details;
    _this.isParent = isParent;
    // Ensure children is an array.
    if (!Array.isArray(_this.children)) {
      _this.children = [_this.children];
    }
    // If this command has no name, it's the "top-level" command. Add a help
    // command and a fallback handler for usage information.
    if (isParent) {
      _this.children = [].concat((0, _toConsumableArray3.default)(_this.children), [_this.createHelpCommand(), _this.createFallbackHandler()]);
    }
    // Keep track of this command's sub-commands for later use.
    _this.subCommands = _this.children.filter(function (c) {
      return c.isCommand;
    });
    // If this command has a name or aliases, create a matching wrapper around
    // children that responds only to that name or an alias.
    if (name || aliases) {
      (function () {
        var items = !aliases ? [] : Array.isArray(aliases) ? aliases : [aliases];
        if (name) {
          items.unshift(name);
        }
        var origChildren = _this.children;
        _this.children = items.map(function (item) {
          return (0, _matcher2.default)({ match: item }, origChildren);
        });
        if (!name) {
          _this.children.push(origChildren);
        }
      })();
    }
    return _this;
  }

  // Does this command have sub-commands?


  (0, _createClass3.default)(CommandMessageHandler, [{
    key: 'hasSubCommands',
    value: function hasSubCommands() {
      return this.subCommands.length > 0;
    }

    // Search for a matching sub-command. If an exact match isn't found, return
    // the closest matching parent command. Returns the matched command, the full
    // name (ie. path) to that command, and whether or not it was an exact match.

  }, {
    key: 'getMatchingSubCommand',
    value: function getMatchingSubCommand(search) {
      var command = this; // eslint-disable-line consistent-this
      var exact = true;
      var prefix = this.isParent ? this.name : '';
      var subCommandNameParts = [];
      if (search) {
        (function () {
          var parts = search.split(/\s+/);

          var _loop = function _loop(_i) {
            var subCommand = command.subCommands.find(function (_ref) {
              var name = _ref.name;

              if (name) {
                // Handle spaces in command names.
                for (var j = _i; j < parts.length; j++) {
                  if (parts.slice(_i, j + 1).join(' ') === name) {
                    _i = j;
                    return true;
                  }
                }
              }
              return false;
            });
            if (!subCommand) {
              exact = false;
              return 'break';
            }
            command = subCommand;
            subCommandNameParts.push(command.name);
            i = _i;
          };

          for (var i = 0; i < parts.length; i++) {
            var _ret3 = _loop(i);

            if (_ret3 === 'break') break;
          }
        })();
      }
      return {
        command: command,
        prefix: prefix,
        exact: exact,
        subCommandName: subCommandNameParts.join(' ')
      };
    }

    // Display usage info for this command, given the specified command name.

  }, {
    key: 'getUsage',
    value: function getUsage(command, prefix) {
      if (!this.name) {
        return false;
      }
      command = formatCommand(prefix, command);
      var usageFormatter = function usageFormatter(details) {
        return formatCommand(command, details);
      };
      // If usage is a function, pass the command to it.
      var usage = typeof this.usage === 'function' ? this.usage(command) :
      // If usage is a string, format it.
      this.usage ? usageFormatter(this.usage) :
      // If usage isn't specified, but the command has sub-commands, format it.
      this.hasSubCommands() ? usageFormatter('<command>') :
      // Otherwise just return the command,
      command;
      return 'Usage: `' + usage + '`';
    }

    // Get help info for this command, given the specified arguments.

  }, {
    key: 'helpInfo',
    value: function helpInfo(search, command, prefix, exact) {
      var helpText = command ? ' for *' + formatCommand(prefix, command) + '*' : prefix ? ' for *' + prefix + '*' : '';
      return [!exact && '_Unknown command *' + formatCommand(prefix, search) + '*, showing help' + helpText + '._', this.description, this.getUsage(command, prefix), this.hasSubCommands() && '*Commands:*', this.subCommands.map(function (c) {
        return '> *' + c.name + '* - ' + c.description;
      }), this.details];
    }

    // Create a top-level "help" command handler that displays help for the
    // closest matching command to what was specified.

  }, {
    key: 'createHelpCommand',
    value: function createHelpCommand() {
      var _this2 = this;

      return createCommand({ // eslint-disable-line no-use-before-define
        name: 'help',
        description: 'Get help for the specified command.',
        usage: '<command>',
        handleMessage: (0, _parser2.default)(function (_ref2) {
          var args = _ref2.args;

          var search = args.join(' ');

          var _getMatchingSubComman = _this2.getMatchingSubCommand(search),
              command = _getMatchingSubComman.command,
              subCommandName = _getMatchingSubComman.subCommandName,
              prefix = _getMatchingSubComman.prefix,
              exact = _getMatchingSubComman.exact;

          return command.helpInfo(search, subCommandName, prefix, exact);
        })
      });
    }

    // Get usage info for this command, given the specified arguments.

  }, {
    key: 'usageInfo',
    value: function usageInfo(message, command, prefix) {
      var isMatch = !message || Boolean(command);
      var usage = isMatch && this.getUsage(command, prefix);
      var help = command = formatCommand(prefix, 'help', command);
      return [!isMatch && 'Unknown command *' + formatCommand(prefix, message) + '*.', usage, (usage ? 'Or try' : 'Try') + ' *' + help + '* for more information.'];
    }

    // Create a top-level "fallback" handler that displays usage info for the
    // closest matching command to what was specified. This handler only runs if
    // every other handler returns false. Ie. no other command matched.

  }, {
    key: 'createFallbackHandler',
    value: function createFallbackHandler() {
      var _this3 = this;

      return function (message) {
        var _getMatchingSubComman2 = _this3.getMatchingSubCommand(message),
            command = _getMatchingSubComman2.command,
            subCommandName = _getMatchingSubComman2.subCommandName,
            prefix = _getMatchingSubComman2.prefix;

        return command.usageInfo(message, subCommandName, prefix);
      };
    }
  }]);
  return CommandMessageHandler;
}(_delegate.DelegatingMessageHandler);

function createCommand() {
  for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    args[_key2] = arguments[_key2];
  }

  return new (Function.prototype.bind.apply(CommandMessageHandler, [null].concat(args)))();
}
