// chatter/dist/message-handler/delegate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DelegatingMessageHandler = undefined;

var _toArray2 = require('babel-runtime/helpers/toArray');

var _toArray3 = _interopRequireDefault(_toArray2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

exports.getHandlers = getHandlers;
exports.default = createDelegate;
exports.composeCreators = composeCreators;

var _processMessage = require('../util/process-message');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Validate these signatures.
// Only options:
//   getHandlers({handleMessage: fn, ...})
// Only children:
//   getHandlers(fn)
//   getHandlers({handleMessage: fn})
//   getHandlers([...])
// Both options and children:
//   getHandlers({...}, fn)
//   getHandlers({...}, {handleMessage: fn})
//   getHandlers({...}, [...])
function getHandlers() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var handlers = arguments[1];

  if (!handlers) {
    handlers = options.handleMessage || options;
  }
  if (!(0, _processMessage.isMessageHandlerOrHandlers)(handlers)) {
    throw new TypeError('Missing required message handler(s).');
  }
  return handlers;
}

var DelegatingMessageHandler = exports.DelegatingMessageHandler = function () {
  function DelegatingMessageHandler(options, children) {
    (0, _classCallCheck3.default)(this, DelegatingMessageHandler);

    this.children = getHandlers(options, children);
  }

  // Iterate over all child handlers, yielding the first non-false result.


  (0, _createClass3.default)(DelegatingMessageHandler, [{
    key: 'handleMessage',
    value: function handleMessage(message) {
      for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }

      return _processMessage.processMessage.apply(undefined, [this.children, message].concat(args));
    }
  }]);
  return DelegatingMessageHandler;
}();

function createDelegate() {
  for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    args[_key2] = arguments[_key2];
  }

  return new (Function.prototype.bind.apply(DelegatingMessageHandler, [null].concat(args)))();
}

// Compose creators that accept a signature like getHandlers() into a single
// creator. All creators receive the same options object.
// Eg:
//   const createMatcherParser = composeCreators(createMatcher, createParser);
//   const fooHandler = createMatcherParser({match: 'foo', parseOptions: {}}, fn);
// Is equivalent to:
//   const fooHandler = createMatcher({match: 'foo'}, createParser({parseOptions: {}}, fn));
function composeCreators() {
  for (var _len3 = arguments.length, creators = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
    creators[_key3] = arguments[_key3];
  }

  if (Array.isArray(creators[0])) {
    creators = creators[0];
  }
  return function composed(options, children) {
    children = getHandlers(options, children);
    function recurse(_ref) {
      var _ref2 = (0, _toArray3.default)(_ref),
          currentHandler = _ref2[0],
          remain = _ref2.slice(1);

      var nextHandler = remain.length > 0 ? recurse(remain) : createDelegate(children);
      return currentHandler(options, nextHandler);
    }
    return recurse(creators);
  };
}
