// chatter/dist/message-handler/matcher.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MatchingMessageHandler = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _get3 = require('babel-runtime/helpers/get');

var _get4 = _interopRequireDefault(_get3);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _toArray2 = require('babel-runtime/helpers/toArray');

var _toArray3 = _interopRequireDefault(_toArray2);

exports.matchStringOrRegex = matchStringOrRegex;
exports.default = createMatcher;

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _delegate = require('./delegate');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Escape special characters that would cause errors if we interpolate them
// into a regex.
function regexEscape(expr) {
  return expr.replace(/[\\^$*+?.()|[\]{}]/g, '\\$&');
}

// If "match" is a String and matches the entire message or matches a space-
// delimited word at the beginning of the message, success. If any text
// remains after the match, return it (with leading spaces stripped) as the
// remainder.
//
// If match is a RegExp and matches the message, return the value of the
// first non-undefined (ie. captured) capture group.
function matchStringOrRegex(match) {
  var message = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

  var re = typeof match === 'string' ? new RegExp('^' + regexEscape(match) + '(?:$|\\s+(.*))', 'i') : match;

  var _ref = message.match(re) || [],
      _ref2 = (0, _toArray3.default)(_ref),
      fullMatch = _ref2[0],
      captures = _ref2.slice(1);

  if (typeof fullMatch !== 'string') {
    return false;
  }
  return captures.find(function (c) {
    return typeof c === 'string';
  }) || '';
}

var MatchingMessageHandler = exports.MatchingMessageHandler = function (_DelegatingMessageHan) {
  (0, _inherits3.default)(MatchingMessageHandler, _DelegatingMessageHan);

  function MatchingMessageHandler() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var children = arguments[1];
    (0, _classCallCheck3.default)(this, MatchingMessageHandler);

    var _this = (0, _possibleConstructorReturn3.default)(this, (MatchingMessageHandler.__proto__ || (0, _getPrototypeOf2.default)(MatchingMessageHandler)).call(this, options, children));

    if (!('match' in options)) {
      throw new TypeError('Missing required "match" option.');
    }
    _this.match = options.match;
    return _this;
  }

  // If match succeeds, pass remainder into child handlers, yielding their
  // result. If no match, yield false.


  (0, _createClass3.default)(MatchingMessageHandler, [{
    key: 'handleMessage',
    value: function handleMessage(message) {
      var _this2 = this;

      for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }

      return _bluebird2.default.try(function () {
        return _this2.doMatch.apply(_this2, [message].concat(args));
      }).then(function (remainder) {
        var _get2;

        if (remainder === false) {
          return false;
        }
        return (_get2 = (0, _get4.default)(MatchingMessageHandler.prototype.__proto__ || (0, _getPrototypeOf2.default)(MatchingMessageHandler.prototype), 'handleMessage', _this2)).call.apply(_get2, [_this2, remainder].concat(args));
      });
    }

    // Attempt to match the message, given the "match" option.

  }, {
    key: 'doMatch',
    value: function doMatch(message) {
      var match = this.match;

      if (typeof match === 'function') {
        for (var _len2 = arguments.length, args = Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
          args[_key2 - 1] = arguments[_key2];
        }

        return match.apply(undefined, [message].concat(args));
      } else if (typeof match === 'string' || match instanceof RegExp) {
        return matchStringOrRegex(match, message);
      }
      throw new TypeError('Invalid "match" option format.');
    }
  }]);
  return MatchingMessageHandler;
}(_delegate.DelegatingMessageHandler);

function createMatcher() {
  for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
    args[_key3] = arguments[_key3];
  }

  return new (Function.prototype.bind.apply(MatchingMessageHandler, [null].concat(args)))();
}
