// chatter/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _bot = require('./bot');

Object.defineProperty(exports, 'Bot', {
  enumerable: true,
  get: function get() {
    return _bot.Bot;
  }
});
Object.defineProperty(exports, 'createBot', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_bot).default;
  }
});

var _slackBot = require('./slack/slack-bot');

Object.defineProperty(exports, 'SlackBot', {
  enumerable: true,
  get: function get() {
    return _slackBot.SlackBot;
  }
});
Object.defineProperty(exports, 'createSlackBot', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_slackBot).default;
  }
});

var _delegate = require('./message-handler/delegate');

Object.defineProperty(exports, 'DelegatingMessageHandler', {
  enumerable: true,
  get: function get() {
    return _delegate.DelegatingMessageHandler;
  }
});
Object.defineProperty(exports, 'createDelegate', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_delegate).default;
  }
});

var _matcher = require('./message-handler/matcher');

Object.defineProperty(exports, 'MatchingMessageHandler', {
  enumerable: true,
  get: function get() {
    return _matcher.MatchingMessageHandler;
  }
});
Object.defineProperty(exports, 'createMatcher', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_matcher).default;
  }
});

var _argsAdjuster = require('./message-handler/args-adjuster');

Object.defineProperty(exports, 'ArgsAdjustingMessageHandler', {
  enumerable: true,
  get: function get() {
    return _argsAdjuster.ArgsAdjustingMessageHandler;
  }
});
Object.defineProperty(exports, 'createArgsAdjuster', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_argsAdjuster).default;
  }
});

var _parser = require('./message-handler/parser');

Object.defineProperty(exports, 'ParsingMessageHandler', {
  enumerable: true,
  get: function get() {
    return _parser.ParsingMessageHandler;
  }
});
Object.defineProperty(exports, 'createParser', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_parser).default;
  }
});

var _conversation = require('./message-handler/conversation');

Object.defineProperty(exports, 'ConversingMessageHandler', {
  enumerable: true,
  get: function get() {
    return _conversation.ConversingMessageHandler;
  }
});
Object.defineProperty(exports, 'createConversation', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_conversation).default;
  }
});

var _command = require('./message-handler/command');

Object.defineProperty(exports, 'CommandMessageHandler', {
  enumerable: true,
  get: function get() {
    return _command.CommandMessageHandler;
  }
});
Object.defineProperty(exports, 'createCommand', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_command).default;
  }
});

var _processMessage = require('./util/process-message');

Object.defineProperty(exports, 'processMessage', {
  enumerable: true,
  get: function get() {
    return _processMessage.processMessage;
  }
});
Object.defineProperty(exports, 'isMessageHandlerOrHandlers', {
  enumerable: true,
  get: function get() {
    return _processMessage.isMessageHandlerOrHandlers;
  }
});

var _argsParser = require('./util/args-parser');

Object.defineProperty(exports, 'parseArgs', {
  enumerable: true,
  get: function get() {
    return _argsParser.parseArgs;
  }
});

var _response = require('./util/response');

Object.defineProperty(exports, 'isMessage', {
  enumerable: true,
  get: function get() {
    return _response.isMessage;
  }
});
Object.defineProperty(exports, 'isArrayOfMessages', {
  enumerable: true,
  get: function get() {
    return _response.isArrayOfMessages;
  }
});
Object.defineProperty(exports, 'normalizeMessage', {
  enumerable: true,
  get: function get() {
    return _response.normalizeMessage;
  }
});
Object.defineProperty(exports, 'normalizeMessages', {
  enumerable: true,
  get: function get() {
    return _response.normalizeMessages;
  }
});
Object.defineProperty(exports, 'normalizeResponse', {
  enumerable: true,
  get: function get() {
    return _response.normalizeResponse;
  }
});

var _queue = require('./util/queue');

Object.defineProperty(exports, 'Queue', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_queue).default;
  }
});
Object.defineProperty(exports, 'composeCreators', {
  enumerable: true,
  get: function get() {
    return _delegate.composeCreators;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
